### DISCUSSION TITLE
enebularでの条件分岐（LINEBot）

### DISCUSSION THREAD

**Discussion by knagaolab** [1ft-seabass/ma-hackathon-qa-and-knowledge-202210#6] [Q&A] [answered]
@1ft-seabass 
ハンズオンで扱った、『lcdp-linebot-simple』をベースに条件分岐を作ろうと思っています。
payloadに格納された文字を"changeノード（メッセージデータ移設）"でpayloadLINEに項目を受け渡しているので、"switchノード"ではpayloadLINEの判定をすればよいと思ったのですが、実際にLINEから行うと分岐も出来なければ返答もありません。
"デバッグノード"の結果を見た感じでは、payloadLINEの文字列に対して分岐をすれば良いと思っているのですが、場所が違うのか、変換などが必要なのか分かりませんでした。ご教示ください。
LCDPではなく、普通のLINEノードを使うだけであれば分岐・置き換えの返答もpayloadだけで上手く出来ました。
※オウム返しや固定文言返答はできているため、基本的な設定は出来ております。
※添付図ではswitchで分岐させていないですが、分岐させても上手くいきません。
![フロー、デバッグ](https://user-images.githubusercontent.com/102508262/199136676-ce35c849-c7db-4d83-80ed-1b0350be4815.JPG)
![各ノード](https://user-images.githubusercontent.com/102508262/199136680-dcef953c-eb29-4710-9da0-43843eb06d27.JPG)
![成功フロー](https://user-images.githubusercontent.com/102508262/199136682-2385e547-d1d6-4ead-bd4f-3883d87745fe.jpg)

**Comment by 1ft-seabass** [comment, marked as answer]
かしこまりました！

一旦ざっくり言うと、LCDPの場合は文字列でなく、JSONデータで来るので、debug ノードでシステムコンソールをチェックして、まずクラウド実行環境に反映。

その後、LINE BOT からアクセスして動かした後、クラウド実行環境のログから確認すると詳細みえます。

のちほど、詳細お伝えしますねー。

**Reply by 1ft-seabass** [reply]
まず、LCDP ノードには LINE BOT からこのような JSON データで来ます。

これは、LINE Messaging API のメッセージオブジェクトです。詳しい仕様は https://developers.line.biz/ja/reference/messaging-api/#wh-text あたりが参考になります。

```js
{
  destination: 'U097dcff4c5a9037af2985d286df70b83',
  events: [
    {
      type: 'message',
      message: { type: 'text', id: '17049248358170', text: 'テスト' },
      webhookEventId: '01GGRSNDVFC2HMXKPNEQSJ2QXF',
      deliveryContext: { isRedelivery: false },
      timestamp: 1667279533428,
      source: { type: 'user', userId: 'Ue8d69a2b51798831eb925fd9938607f8' },
      replyToken: 'c57f2437e594411abf01fd0a7835930d',
      mode: 'active'
    }
  ]
}
```

直前の「メッセージデータ移設」で msg.payload の中身は msg.payloadLINE に移設しているので、そのあとの switch ノードでは msg.payloadLINE.events[0].message.text でメッセージ内容を捉えることができます。

以下のように change ノードの中身を設定すると、たとえば「テスト」という文言で分岐ができます。

![image](https://user-images.githubusercontent.com/2148840/199165140-5b54857f-224b-481a-a424-8457b34b9811.png)

- ルール 1 つ目
  - プロパティ
    - msg.payloadLINE.events[0].message.text
  - `==` で `文字列「テスト」`
- ルール 2 つ目
  - プロパティ
    - その他

以降はこんな分岐です。

![image](https://user-images.githubusercontent.com/2148840/199164692-b4ec45dd-a7bd-43a7-92f8-7ca0072efbad.png)

フローを保存した後、再度クラウド実行環境にデプロイします。

![image](https://user-images.githubusercontent.com/2148840/199165507-e876a82c-6acd-4ddb-bf12-7ee010ed2c40.png)

動かしてみると、テストの文言の時だけ気の抜けた「テストする～？」ときて、それ以外は今まで通りオウム返しします。

**Reply by 1ft-seabass** [reply]
@knagaolab より深く開発していくときにログでデータ見たいってことになると思うので、以下の「LCDP ノードを使ったフローのデバッグ方法」のドキュメントまとめたのでよろしければご参考くださいー。さっきのような LINE BOT からのデータを見るときも、このやり方が使えます。

https://1ft-seabass.github.io/ma-enebular-handson-202210/70-01-tips-lcdp-debug.html

**Reply by knagaolab** [reply]
ありがとうございます。出来ました！
LCDPではJSON形式データに対して、どこに格納されているかコンソールで確認⇒正しい場所を指定して取得するということで理解しました。
これをベースに他APIとも接続などをすることになると思いますので、色々試してみます！
<img width="411" alt="Screenshot_20221101-150255" src="https://user-images.githubusercontent.com/102508262/199178214-973383cb-28f4-479a-8f49-320603af3e86.png">

**Reply by 1ft-seabass** [reply]
@knagaolab うおおおお！やったー。よかったです！

早速ありがとうございました～。

**Reply by 1ft-seabass** [reply]
キャプチャのありがとう感が細かくて好きですw
